src/order/minimal.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.antichain
+
+/-!
+# Minimal/maximal elements of a set
+
+This file defines minimal and maximal of a set with respect to an arbitrary relation.
+
+## Main declarations
+
+* `maximals r s`: Maximal elements of `s` with respect to `r`.
+* `minimals r s`: Minimal elements of `s` with respect to `r`.
+
+## TODO
+
+Do we need a `finset` version?
+-/
+
+open function set
+
+variables {α : Type*} (r r₁ r₂ : α → α → Prop) (s t : set α) (a : α)
+
+/-- Turns a set into an antichain by keeping only the "maximal" elements. -/
+def maximals : set α := {a ∈ s | ∀ ⦃b⦄, b ∈ s → r a b → a = b}
+
+/-- Turns a set into an antichain by keeping only the "maximal" elements. -/
+def minimals : set α := {a ∈ s | ∀ ⦃b⦄, b ∈ s → r b a → a = b}
+
+lemma maximals_subset : maximals r s ⊆ s := sep_subset _ _
+lemma minimals_subset : minimals r s ⊆ s := sep_subset _ _
+
+@[simp] lemma maximals_empty : maximals r ∅ = ∅ := sep_empty _
+@[simp] lemma minimals_empty : minimals r ∅ = ∅ := sep_empty _
+
+@[simp] lemma maximals_singleton : maximals r {a} = {a} :=
+(maximals_subset _ _).antisymm $ singleton_subset_iff.2 $ ⟨rfl, λ b hb _, hb.symm⟩
+
+@[simp] lemma minimals_singleton : minimals r {a} = {a} := maximals_singleton _ _
+
+lemma maximals_swap : maximals (swap r) s = minimals r s := rfl
+lemma minimals_swap : minimals (swap r) s = maximals r s := rfl
```

Comment:
Should these be `@[simp]`?

Reply:
No, for the same reason as for `bipartite_below`/`bipartite_above`. Deceivingly, the LHS is not simpler than the RHS.